File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/playlist/specs/GivenAStandardPreparedPlaylistProvider/WithAStatefulPlaybackHandler/ThatIsPlaying/AndThenPaused/WhenResumingPlayback.java
package com.lasthopesoftware.bluewater.client.playback.playlist.specs.GivenAStandardPreparedPlaylistProvider.WithAStatefulPlaybackHandler.ThatIsPlaying.AndThenPaused;

import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue;
import com.lasthopesoftware.bluewater.client.playback.file.NoTransformVolumeManager;
import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.FakeBufferingPlaybackHandler;
import com.lasthopesoftware.bluewater.client.playback.playlist.IPlaylistPlayer;
import com.lasthopesoftware.bluewater.client.playback.playlist.PlaylistPlayer;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import io.reactivex.Observable;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenResumingPlayback {

	private FakeBufferingPlaybackHandler playbackHandler;

	@Before
	public void before() throws ExecutionException, InterruptedException {
		playbackHandler = new FakeBufferingPlaybackHandler();

		final Promise<PositionedPlayableFile> positionedPlaybackHandlerContainer =
			new Promise<>(new PositionedPlayableFile(
				0,
				playbackHandler,
				new NoTransformVolumeManager(),
				new ServiceFile(1)));

		final PreparedPlayableFileQueue preparedPlaybackFileQueue = mock(PreparedPlayableFileQueue.class);
		when(preparedPlaybackFileQueue.promiseNextPreparedPlaybackFile(0))
			.thenReturn(positionedPlaybackHandlerContainer);

		final IPlaylistPlayer playlistPlayback = new PlaylistPlayer(preparedPlaybackFileQueue, 0);

		Observable.create(playlistPlayback).subscribe();

		new FuturePromise<>(playlistPlayback.pause()).get();

		new FuturePromise<>(playlistPlayback.resume()).get();
	}

	@Test
	public void thenPlaybackIsResumed() {
		assertThat(playbackHandler.isPlaying()).isTrue();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/playlist/GivenAStandardPreparedPlaylistProvider/WithAStatefulPlaybackHandler/ThatIsPlaying/AndThenPaused/WhenResumingPlayback.kt
package com.lasthopesoftware.bluewater.client.playback.playlist.GivenAStandardPreparedPlaylistProvider.WithAStatefulPlaybackHandler.ThatIsPlaying.AndThenPaused

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.SupplyQueuedPreparedFiles
import com.lasthopesoftware.bluewater.client.playback.file.NoTransformVolumeManager
import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayingFile
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakeBufferingPlaybackHandler
import com.lasthopesoftware.bluewater.client.playback.playlist.PlaylistPlayer
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.joda.time.Duration
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

class WhenResumingPlayback {

	private val playbackHandler = FakeBufferingPlaybackHandler()
	private var playingFile: PositionedPlayingFile? = null

	@BeforeAll
	fun act() {
		val positionedPlaybackHandlerContainer = Promise(PositionedPlayableFile(
			0,
			playbackHandler,
			NoTransformVolumeManager(),
			ServiceFile("1")))
		val preparedPlaybackFileQueue = mockk<SupplyQueuedPreparedFiles>().apply {
			every { promiseNextPreparedPlaybackFile(Duration.ZERO) } returns positionedPlaybackHandlerContainer
		}
		val playlistPlayback = PlaylistPlayer(preparedPlaybackFileQueue, Duration.ZERO)
		playlistPlayback.resume().toExpiringFuture().get()
		playlistPlayback.pause().toExpiringFuture().get()
		playingFile = playlistPlayback.resume().toExpiringFuture().get()
	}

	@Test
	fun `then playback is resumed`() {
		assertThat(playbackHandler.isPlaying).isTrue
	}

	@Test
	fun `then the playlist position is correct`() {
		assertThat(playingFile?.playlistPosition).isEqualTo(0)
	}

	@Test
	fun `then the playing file is correct`() {
		assertThat(playingFile?.serviceFile).isEqualTo(ServiceFile("1"))
	}
}
